source/adaptive_load/adaptive_load_controller_impl.cc

```diff
@@ -0,0 +1,444 @@
+#include <chrono>
+
+#include "envoy/common/exception.h"
+#include "envoy/config/core/v3/base.pb.h"
+
+#include "nighthawk/adaptive_load/adaptive_load_controller.h"
+#include "nighthawk/adaptive_load/metrics_plugin.h"
+#include "nighthawk/adaptive_load/scoring_function.h"
+#include "nighthawk/adaptive_load/step_controller.h"
+
+#include "external/envoy/source/common/common/logger.h"
+#include "external/envoy/source/common/common/statusor.h"
+#include "external/envoy/source/common/protobuf/protobuf.h"
+
+#include "api/adaptive_load/adaptive_load.pb.h"
+#include "api/adaptive_load/metric_spec.pb.h"
+#include "api/client/options.pb.h"
+#include "api/client/output.pb.h"
+#include "api/client/service.grpc.pb.h"
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/status/status.h"
+#include "absl/strings/str_join.h"
+#include "adaptive_load/metrics_plugin_impl.h"
+#include "adaptive_load/plugin_loader.h"
+
+namespace Nighthawk {
+
+namespace {
+
+using namespace Envoy; // for Envoy::Logger
+
+using nighthawk::adaptive_load::AdaptiveLoadSessionOutput;
+using nighthawk::adaptive_load::AdaptiveLoadSessionSpec;
+using nighthawk::adaptive_load::BenchmarkResult;
+using nighthawk::adaptive_load::MetricEvaluation;
+using nighthawk::adaptive_load::MetricSpec;
+using nighthawk::adaptive_load::MetricSpecWithThreshold;
+using nighthawk::adaptive_load::ThresholdSpec;
+
+/**
+ * Runs a single benchmark using a Nighthawk Service.
+ *
+ * @param nighthawk_service_stub Nighthawk Service gRPC stub.
+ * @param command_line_options Nighthawk Service benchmark request proto generated by the
+ * StepController, without the duration set.
+ * @param duration Duration to insert into the benchmark request.
+ *
+ * @return ExecutionResponse If we reached the Nighthawk Service, this is the raw ExecutionResponse
+ * proto, containing the benchmark data or possibly an error message from Nighthawk Service; if we
+ * had trouble communicating with the Nighthawk Service, we insert the details into the
+ * |error_detail| of this proto.
+ */
+nighthawk::client::ExecutionResponse PerformNighthawkBenchmark(
```

Comment:
Reading this file again, and probably for after this PR has landed / a nice to have, I wonder how hard it would be to extract an interface from this that could (hypothetically) support both CLI and gRPC based flows.
Then this implementation would be the first one, and serve as the canonical example of how to make the rubber hit the road with respect to the adaptive load controller framework. 

If this is desireable, one thing that I think that would be good to consider is naming. For example, maybe this class could then better be called something like `AdaptiveNighthawkGrpcLoadController`?  Also, some of the class method names mention "Nighthawk". We'd want to be less specific, as some of those might end up being overrides of the interface spec.

Reply:
There's actually not a class here, only a bare public function and helper functions in the private namespace. This might all have to be refactored if we need to open it to overriding some of the parts, including rethinking the signatures of the functions that rely on Nighthawk-specific protos. Opened #485 to consider this area.

Reply:
Thanks for providing context and filing that issue. Resolving this here.